### PR TITLE
[NT-930] Add Project Creator Details Query and VM outputs

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Adds a query for us to fetch some additional project creator details from GraphQL when a project page loads.

**Note:** This emits what we expect to need to configure the cell that will display this data, that is the `ProjectCreatorDetailsEnvelope` result and an `isLoading` boolean to indicate whether we are still fetching the results.

# 🤔 Why

As part of an experiment that we are running we would like to display some additional data to surface trust indicators to users. These values are not available from the v1 API so we are fetching them from GraphQL.

# 🛠 How

- Added `ProjectCreatorDetailsEnvelope`.
- Added `fetchProjectCreatorDetails` to `ServiceType`.
- Added some types to `ProjectPamphletContentViewModel` to improve the frequently changing interface of its datasource loading output.
- Updated `GraphSchema`.
- Added tests.

# ✅ Acceptance criteria

- [ ] Upon a project page loading observe that `(nil, true)` is emitted to put the cell in an initial loading state.
- [ ] Once the GraphQL request successfully completes observe that `(result, false)` is emitted.
- [ ] If the request fails observe that `(nil, false)` is emitted.

### CODE REVIEW

File: Kickstarter-iOS/DataSources/ProjectPamphletContentDataSource.swift [hunk 1]
```diff
@@ -39,7 +39,7 @@ internal final class ProjectPamphletContentDataSource: ValueCellDataSource {
     }
 
     self.set(
-      values: [projectAndRefTag],
+      values: [(project, refTag)],
```
Comment (justinswart) [inline Kickstarter-iOS/DataSources/ProjectPamphletContentDataSource.swift]: We will update `ProjectPamphletMainCell` to be configured with all three values (or just `ProjectPamphletContentData`).

File: KsApi/MockService.swift [hunk 2]
```diff
@@ -93,11 +93,13 @@
...
     fileprivate let fetchProjectError: ErrorEnvelope?
 
-    fileprivate let fetchProjectNotificationsResponse: [ProjectNotification]
-
     fileprivate let fetchProjectsResponse: [Project]?
     fileprivate let fetchProjectsError: ErrorEnvelope?
 
+    fileprivate let fetchProjectCreatorDetailsResult: Result<ProjectCreatorDetailsEnvelope, GraphError>?
+
+    fileprivate let fetchProjectNotificationsResponse: [ProjectNotification]
```
Comment (justinswart) [inline KsApi/MockService.swift]: Alphabetized this so it moved around in a few places. But I now realize a bunch of things are still out of place so I might revert this change and rather fix the ordering separately.

File: Library/ViewModels/ProjectPamphletContentViewModel.swift [hunk 3]
```diff
@@ -58,8 +61,27 @@ public final class ProjectPamphletContentViewModel: ProjectPamphletContentViewMo
...
+    let projectCreatorDetails = project
+      .take(first: 1)
+      .map(\.slug)
+      .switchMap { slug in
+        AppEnvironment.current.apiService
+          .fetchProjectCreatorDetails(query: projectCreatorDetailsQuery(withSlug: slug))
+          .ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)
+          .map { result in (result, false) }
+          .prefix(value: (nil, true))
+          .demoteErrors(replaceErrorWith: (nil, false))
```
Comment (justinswart) [inline Library/ViewModels/ProjectPamphletContentViewModel.swift]: If this errors out we just return `nil` and tell the cell to stop loading. We will do additional work to not even activate the experiment if we don't have this data.

File: KsApi/Service.swift [hunk 4]
```diff
@@ -306,6 +302,14 @@ public struct Service: ServiceType {
     return requestPagination(paginationUrl)
   }
 
+  public func fetchProjectCreatorDetails(query: NonEmptySet<Query>) -> SignalProducer<ProjectCreatorDetailsEnvelope, GraphError> {
```
Comment (nativeksr) [inline KsApi/Service.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/kickstarter/ios-oss/blob/a8f032cddfba991aba476d145bc2cbcf057daea7/KsApi/Service.swift#L305"></span>Line should be 110 characters or less: currently 130 characters<br />
<code>line_length</code> <code>Service.swift:305</code></td>
    </tr>
  </tbody>
</table>

File: Library/ViewModels/ProjectPamphletContentViewModelTests.swift [hunk 5]
```diff
@@ -390,4 +399,144 @@ final class ProjectPamphletContentViewModelTests: TestCase {
...
+      // navigator controller.
+      self.vm.inputs.viewWillAppear(animated: false)
+      self.vm.inputs.viewDidAppear(animated: false)
+
+      self.loadProjectIntoDataSourceCreatorDetailsIsLoading.assertValues([true], "Starts loading.")
+      self.loadProjectIntoDataSourceCreatorDetailsEnvelope.assertValues([nil], "Emits nil as view appears")
+      self.loadProjectIntoDataSourceProject
+        .assertValues([project], "Load the full project into the data source.")
+      self.loadProjectIntoDataSourceRefTag.assertValues([.discovery], "Load the refTag into the data source.")
+      self.loadMinimalProjectIntoDataSource.assertValues([], "Do not load the minimal version of the project.")
```
Comment (nativeksr) [inline Library/ViewModels/ProjectPamphletContentViewModelTests.swift]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/kickstarter/ios-oss/blob/a8f032cddfba991aba476d145bc2cbcf057daea7/Library/ViewModels/ProjectPamphletContentViewModelTests.swift#L501"></span>Line should be 110 characters or less: currently 111 characters<br />
<code>line_length</code> <code>ProjectPamphletContentViewModelTests.swift:501</code></td>
    </tr>
  </tbody>
</table>

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: KsApi/GraphSchema.swift [hunk 6]
```diff
@@ -192,17 +193,18 @@ public enum Query {
     }
   }
 
-  public enum User {
+  public indirect enum User {
```
Comment (Scollaco) [inline KsApi/GraphSchema.swift]: Is there a reason why we are enabling indirection for all cases?

File: Library/ViewModels/ProjectPamphletContentViewModelTests.swift [hunk 7]
```diff
@@ -6,30 +6,39 @@ import ReactiveSwift
 import XCTest
 
 final class ProjectPamphletContentViewModelTests: TestCase {
-  fileprivate let vm: ProjectPamphletContentViewModelType = ProjectPamphletContentViewModel()
+  fileprivate var vm: ProjectPamphletContentViewModelType!
```
Comment (Scollaco) [inline Library/ViewModels/ProjectPamphletContentViewModelTests.swift]: Is this change necessary? 🤔

File: KsApi/GraphSchema.swift (hunk 6 above)
Comment (justinswart) [reply KsApi/GraphSchema.swift]: I think this is because of the recursive relationship between `Project` <-> `User`. Without this it gives this error:
image

File: Library/ViewModels/ProjectPamphletContentViewModelTests.swift (hunk 7 above)
Comment (justinswart) [reply Library/ViewModels/ProjectPamphletContentViewModelTests.swift]: Nope! Sometimes it is for the scheduler to work correctly but in this case it wasn't. Thanks 👍
